client/history/redirector.go

```diff
@@ -0,0 +1,146 @@
+	// instance owns the shard.
+	operationTarget struct {
+		shardID int32
+		address rpcAddress
+	}
+
+	// A redirector executes a client operation against a history instance.
+	// If the operation is intended for the owner of a shard, and the request
+	// returns a shard ownership lost error with a hint for a new shard owner,
+	// the redirector will retry the request to the new owner.
+	redirector interface {
+		clientForTarget(operationTarget) (historyservice.HistoryServiceClient, error)
+		execute(context.Context, operationTarget, clientOperation) error
+	}
+
+	clientOperation func(ctx context.Context, client historyservice.HistoryServiceClient) error
+
+	basicRedirector struct {
+		connections            *clientConnections
+		historyServiceResolver membership.ServiceResolver
+	}
+)
+
+func (t operationTarget) validate() error {
+	if t.shardID < 0 {
+		return serviceerror.NewInvalidArgument(fmt.Sprintf("Invalid ShardID: %d", t.shardID))
+	}
+	if t.shardID == 0 {
+		if len(t.address) == 0 {
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("Invalid client target"))
+		}
+	} else {
+		if len(t.address) != 0 {
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("Invalid client target"))
+		}
+	}
+	return nil
+}
+
+func shardLookup(resolver membership.ServiceResolver, shardID int32) (rpcAddress, error) {
+	hostInfo, err := resolver.Lookup(convert.Int32ToString(shardID))
+	if err != nil {
+		return "", err
+	}
+	return rpcAddress(hostInfo.GetAddress()), nil
+}
+
+func newBasicRedirector(
+	connections *clientConnections,
+	historyServiceResolver membership.ServiceResolver,
+) *basicRedirector {
+	return &basicRedirector{
+		connections:            connections,
+		historyServiceResolver: historyServiceResolver,
+	}
+}
+
+func (r *basicRedirector) clientForTarget(target operationTarget) (historyservice.HistoryServiceClient, error) {
+	if err := target.validate(); err != nil {
+		return nil, err
+	}
+	if target.shardID == 0 {
+		connection := r.connections.getOrCreateClientConn(target.address)
+		return connection.historyClient, nil
+	}
+	address, err := shardLookup(r.historyServiceResolver, target.shardID)
+	if err != nil {
+		return nil, err
+	}
+	clientConn := r.connections.getOrCreateClientConn(address)
+	return clientConn.historyClient, nil
+}
+
+func (r *basicRedirector) execute(ctx context.Context, target operationTarget, op clientOperation) error {
+	if err := target.validate(); err != nil {
+		return err
+	}
+	if target.shardID == 0 {
+		connection := r.connections.getOrCreateClientConn(target.address)
+		return op(ctx, connection.historyClient)
+	}
+	address, err := shardLookup(r.historyServiceResolver, target.shardID)
+	if err != nil {
+		return err
+	}
+	return r.redirectLoop(ctx, address, op)
+}
+
+func (r *basicRedirector) redirectLoop(ctx context.Context, address rpcAddress, op clientOperation) error {
+	for {
+		if err := common.IsValidContext(ctx); err != nil {
```

Comment:
Can you wrap this error with something? Getting naked "context deadline exceeded" errors can be hard to troubleshoot. I think including the number of redirects could be useful information to include for instance.